### ISSUE TITLE
Suggestions for new combi mode

### ISSUE THREAD

**Issue by ThomasGaertner** [Courseplay/courseplay#3282]
I want to start with the refactoring of mode 2\3 combi\overloader.
for that I'm collecting oppinions for certain situations. I want to ask you to tell your oppinion to the following situations:
- tractor next to combine, combine is empty -> where should we park the tractor ?
- chopper is turning on  field edge, how do you imagine the turn maneuver of the tractor ?
- who uses "search combine on field xx" already ?
- any other remarks what should be changed ? now or never :-)

**Comment by sootysax** [comment]
1.  you can leave the tractor right where it is, unless it's in the crop
2. the tractor should turn in the empty headlands of the field, without going outside the field edge - after the combine has finished its turn of course.  perhaps if the combine is not excessively full, the tractor can make its turn more toward the middle of the field and line itself up on the crop edge, and wait for the combine to pass it.. rather than trying to always get behind the turning combine and crashing into a fence
3. i'm using the search on field function already

**Comment by alpha117** [comment]
Agree with Sootysax, I'm also already using the search on field function

**Comment by sootysax** [comment]
I think the combine should set itself to 'stop while unloading' if the pipe is in the fruit

**Comment by alpha117** [comment]
Would it be feasible to only unload when pipe is on opposite side to fruit?

**Comment by sootysax** [comment]
No, because sometimes the combine will reach 100% when the pipe is in the fruit

**Comment by zed636** [comment]
If the pipe is in fruit could you pull the harvester back as the minute 5:00 of the video below?

**Comment by StefanBraun2001** [comment]
1-3 agreeing with sootysax.
One thing: In case of choppers or sugar cane harvester the harvester/chopper countine harvesting while tractor is still turning (driving backwards to field edge in most cases) Is it possible to let the harvester wait until tractor is finished turning?

**Comment by alpha117** [comment]
I like that idea

![image](https://user-images.githubusercontent.com/30005040/53246503-db1c9580-36a8-11e9-8cd7-a9ce6d4dcefb.png)

**Comment by Jimmy-the-Geek** [comment]
I'm with sootysax, however, I do like alpha117's idea of pulling the
combine back to unload when the pipe is in fruit. That's much more
realistic than the chaser running through the crop.

**Comment by zed636** [comment]
For the turning point with the forage harvester, could you make a 45-degrees turn at the opposite side of the row where you should enter and the make a confortable turn to both, the tractor with trailer and the chopper? E.g. https://youtu.be/o2TPkjcV3bM

**Comment by pvaiko** [comment]
I'd definitely leave the tractor where it is. 

And love the idea of pulling the harvester back when the pipe is in the fruit.

**Comment by Bobster82** [comment]
* tractor next to combine, combine is empty -> where should we park the tractor ?
     Right where he is when combine is empty, if it is the headland, park behind the combine (as what it is now) if it is in the fruit, park next to the fruit.
* chopper is turning on  field edge, how do you imagine the turn maneuver of the tractor ?
     Turn away from the chopper (opposite), make turn and get next to the chopper again
* who uses "search combine on field xx" already ?
     I do, i like it, makes it easier to find more combines in a field when u are parked outside the field.
* any other remarks what should be changed ? now or never :-)
     - when combine is unloading into a trailer and want to make a end field turn, stop, empty complete before turning and when empty(or trailer full) continue without waiting for tracktor to park (or whatever)
     - slider to change the convoy distance
     - Better and earlier traffic collision check

Just quick things that pop up now :P

**Comment by pseikoNL** [comment]
When pipe is in fruit then pull back the harvester and offset it to the opposite side so there’s room for the tractor to go under pipe and don’t drive into the fruit, after unloading the tractor can pull back a few meters, waits until harvester continues to harvest and then park on the spot where the harvester stands when unloading.

If pipe during first headland is outside the field boundaries, and there are collisions outside the field boundaries like fenches etc., then pullback the harvester in a quarter of a circle so there is hopefully enough room to park the tractor under the pipe and don’t collide into something, after unloading the tractor will pull back enough to give the harvester room to leave, and then park on the spot of the harvester where it stands when continueing the harvest (not the point of unloading)

**Comment by mehdinasef** [comment]
i'd like the tractor to drive behind the chopper in some situations, for example when doing the first headland (to avoide geting stuck). As in minute 2:00 of this video:

https://www.youtube.com/watch?v=bpbFr4xacZo&t=54s

**Comment by GeorgeEarslight** [comment]
@ThomasGaertner Thanks for asking the community's opinion.

* chopper is turning on  field edge, how do you imagine the turn maneuver of the tractor ?
  - Don't follow the havester on turn maneuvers at all. Prefferably stop both machines before the turn.
  - Follow normally on headlands but make sure that pathfinding never takes the tractor outside the field, or inside fruit. 
  - Using reverse to unload could also work, but in my opinion is realism breaking. There are valid reasons of why many farms explicitly forbid harvester reverse maneuvers. 
* who uses "search combine on field xx" already ?
  - I do, no problems so far F15 and F18 Marwell.
* any other remarks what should be changed ? now or never :-)
  - Allow the tractor to use the combine's already harvested route to return to the field exit, no need for pathfinding that way.
* tractor next to combine, combine is empty -> where should we park the tractor ?
  - OK, this is a long one. It's a specific case of UP-DOWN course unloading.
  - Example harvester has been unloaded at  it's course point `iXa=100`.  
  - Calculate fill rate and predict next unload `iXb`.  
  - Use the tractor's horizontal offset and get the line equations `y=mx+c` of lines parallel to the harvester when on `iXa` and `iXb`. 
  - Get the coordinates of the point `iXc `on line a which is perpendicular to `iXb`. 
  - Match the predicted direction of the harvester on `iXb`, 
  - Move back `k` meters equal to tractor's turn radius and you have `iXd`, **which is where you park**
  - Picture attached with predicted unload course `c`

![git002](https://user-images.githubusercontent.com/30686125/53251295-fdbca780-36c4-11e9-8746-64a709b06719.png)

**Comment by pvaiko** [comment]
@GeorgeEarslight that's too dependent on the up/down row pattern. We have other patterns to take care of as well.

I still think the tractor should just stay where it is, otherwise it will get too complicated.

What we could do @ThomasGaertner is to add more intelligence to the combine for example if it thinks that it'll reach the set full level in the next row and the pipe would be in the fruit on that row then  it stops and calls the tractor instead of making a turn.

**Comment by sootysax** [comment]
We should also discuss the tractor's behaviour when unloading harvesters that are threshing the headlands.  Particularly the outer ring of headlands.   This is when the tractor most often crashes into fencing trying to turn behind the combine.   The tractor should probably always travel along the headland in the direction of the course.

**Comment by ChrisASnyder** [comment]
Having more control over how tractors enter, exit and navigate fields would be helpful for fields with an odd shape or with islands. 
A possible approach: 
	- Create a safe route for the tractor to follow with way points along the way. Use the way point closest to the harvester as departure point. 
	- When the tractor needs to leave the field, head to the closest waypoint and follow the path.
	- Add a wait point as the location for tractors to wait and search for harvesters when entering.
	
Lots of good ideas in this thread.

**Comment by sootysax** [comment]
> Having more control over how tractors enter, exit and navigate fields would be helpful for fields with an odd shape or with islands.

You have control over recording your unload course..   you should start /end your unload course at a place reachable from any point on the field (in the case of an L-shaped field, the middle of the field).  This goes for any unload/refill course.

**Comment by Mannelig83** [comment]
I would like to see a unique naming pattern if you use more then 1 harvester of same type on the same field to assign the tractors correct
and 
if you have 2 harvesters and 4 tractors on the field (like i have actually) then you get a lot of tractor traffic on the field from harvesters to unloading Course and sometimes a crash of machines.
So it would make sense in my opinion that you ceep clear a area around a automatic set waypoint on the field.

Greets
Markus

**Comment by KenStumich** [comment]
* tractor next to combine, combine is empty -> where should we park the tractor ?
leave it , until there is a reason to move.

* who uses "search combine on field xx" already ?
i do, becouse there is not always a place on the field to park the the traktor while waiting.

* any other remarks what should be changed ? now or never :-)

combine fist row clockwise.cant make it to the end to unload. 
suggestion > backup a little an cut a pocket in the lane to the right to get the needet space to unload without to leave the field and not to drive into the fruit while unloading.

**Comment by GeorgeEarslight** [comment]
> @GeorgeEarslight that's too dependent on the up/down row pattern. We have other patterns to take care of as well.
> 
> I still think the tractor should just stay where it is, otherwise it will get too complicated.
> 
> What we could do @ThomasGaertner is to add more intelligence to the combine for example if it thinks that it'll reach the set full level in the next row and the pipe would be in the fruit on that row then it stops and calls the tractor instead of making a turn.

@pvaiko Staying put would also work fine and would obviously be quicker to implement.
I still think that each pattern could be further optimized in the future though, calling it's own parking function. I could write some skeleton code, but I'm only fluent in C.

**Comment by KenStumich** [comment]
suggestion > headland turn without leaving the field

![fsscreen_2019_02_22_17_42_58](https://user-images.githubusercontent.com/28400454/53257660-0617e100-36cb-11e9-8643-fb5db31b5d25.png)
![fsscreen_2019_02_22_17_43_26](https://user-images.githubusercontent.com/28400454/53257690-10d27600-36cb-11e9-872b-2885d477cb6d.png)
![fsscreen_2019_02_22_17_44_53](https://user-images.githubusercontent.com/28400454/53257701-15972a00-36cb-11e9-9e87-237a1820419c.png)
![fsscreen_2019_02_22_17_45_08](https://user-images.githubusercontent.com/28400454/53257707-192ab100-36cb-11e9-911d-230d125264d9.png)
![fsscreen_2019_02_22_17_45_26](https://user-images.githubusercontent.com/28400454/53257716-1f209200-36cb-11e9-9e4c-a90d27a023e9.png)

you always get sharp corners and there is no problem with fences or something else. 

im a great fan of the option "HEADLAND_MODE_TWO_SIDE" cause this is what i think is the most natural way to work.

**Comment by pvaiko** [comment]
> I'm only fluent in C

@GeorgeEarslight what a lame excuse :)

**Comment by KenStumich** [comment]
* chopper is turning on  field edge, how do you imagine the turn maneuver of the tractor ?


first row  behind the cutter like this
![fsscreen_2019_02_22_18_27_12](https://user-images.githubusercontent.com/28400454/53260805-b6d5ae80-36d2-11e9-9e02-10e62cef0f67.png)

normal 1 turn at the moment ( turn outside the field )
![fsscreen_2019_02_22_18_47_02](https://user-images.githubusercontent.com/28400454/53260913-0320ee80-36d3-11e9-8130-6191408aa9ba.png)

expectation

back up a little with both units and turn inside the field.  (the traktor will catch the caff just fine behind the cutter , while the cutter is turning.
![fsscreen_2019_02_22_18_56_15](https://user-images.githubusercontent.com/28400454/53261430-4a5baf00-36d4-11e9-8445-b300cb6cd38a.png)
![fsscreen_2019_02_22_18_56_53](https://user-images.githubusercontent.com/28400454/53261443-547dad80-36d4-11e9-8d35-0259c515be53.png)

there is plenty of room to turn behind the cutter, when the cutter-turn is done. 

also. if the tractor is almost full a second tractor can go beside the fist one and can continue while the first one only needs to stop .
no need to get complicate with something like a cutter. there are no big trailer on the field.  you need something small with a narrow turn radius.

we are talking about cp helper. so use some helper, not only 1 (big fat trailer) to get the worke done.

**Comment by sootysax** [comment]
Pathfinding tractors should also avoid collisions with other vehicles, fences, poles, bales.

**Comment by KenStumich** [comment]
just to be clear , the meaning of pathfinding is to avoid the fruit on the field ?
(not native english speaker)

**Comment by KenStumich** [comment]
if we are brainstorming here , is there a set of rules we need to archive ? some possible suggestions may be nice, but how often is this really usefull ? 
if we (you) are plannig to revisite the code , what is the goal at all ? are we trying to be realistic, or are we going to archive the easy way to do somthing ? 
if we define some major goals we are able to discuss some pros and cons and if they are possible. 

eg. why are we need to discuss where to park a tractor  when the harvester is empty ? normaly the harvester is between the fruit and the tractor . no reason to move it at all. 
can we archive this allways ? yes we can. 
so why discuss this ? 

if we archive some ground rules, there is no need to discuss some problems we dont have if we are going the easy way in the beginning. 

TLDR: 
 what is the ultimate gole and what are the rules ?

**Comment by pAnFarmer** [comment]
Hi. Well this is a quite interesting thread.
Like many wrote before, I would leave the tractor waiting on the unloadpoint after harvester is emptied. Like I suggested already in FS17. I do not see reasons why the tractor should move anywhere after being loaded (except if to go unload of course). Like this you've got hell number of possible crashes in/outside the field banned, since there is no movement of the tractor. Many times tractors tend to go to any waitingpoints (where do they get those) through houses and trees. Just let it wait where it is.

Many good suggestions were made on the issue of unloading with pipe in crop. Indeed, it would be nicer not to drive through the crop. Some manoeuvres or stopping of the harvester and call tractor before turning pipe in the crop is the nicest solution I guess (no unnecessary movements = prevent crashes). If on first headland, the solution of making a pouch in the 2nd headland to allow tractor to pull next to harvester without leaving field (no crashes with decorations...). In general all the machines should stay on the field to make their manoeuvers.

I use the search harvester on field too, one of the nicest features added lately  I think.

New things: 
-In FS17 I had lots of problems with tractors making circles to get on the first point of its unload-route after being full.  Even the  go on straight to first point didn't help. Sometimes on the transition from headlands to up and down pattern as well (this happened on the merged fiels 21+22 on Felsbrun at the northeast sharp corner with rsm 161). Maybe it cannot be reproduced and thus fixed. My suggesstion would be: could you find a way to make cp detect such behaviour (driving circles for hours) and inform with a warning box?
-A better crash prevention on fields would be nice. Sometimes the tractors cut the waypoints in front of a harvester to get behind it to load and the harvester will win the crashbattle. 

Thx for your great work. I hope you find some smart solutions to make the community happy once more.

pa

**Comment by pops64** [comment]
Has someone that was trying to improve the old one. I know leaving the tipper where it is doesn't work for field edge cases as this interferes with the combines turn manuver and depending on the map it could also leave it parked in a roadway.  There also could be issues with those when using multiple combines. Also having the tipper move gives people that have other tools working the field that extra space

My thoughts are to fix the annoying behavior the 17 version had.

Cutting the combines off when procdeing to the unload course.

Pathfinding on 100 of the time like mode 4/6 was at the end

Smarter detection of when to use the reverse turn manuver.

Know if it it will fill when the pipe is over the fruit and call the tipper early.

Know if it is about to cause a traffic issue with other tools working the field and repark it self.at the combines current location.

Scrap the whole headland parking manuver. And make it smarter. This was usually the parking manuver that messed up most of the time. Parked it the way of the combines that just moved onto up down rows. Parked out side the field. Parked in fruit. etc.

For choppers Back up a little to remove the TC with the chopper. Then see if it can make a circle turn while staying in the field and back up has much has needed. Has the current system is dumb and just backs a x amount. Then have the chopper wait for the turn to complete before proceeding. Also use the choppers current location as an indicator of where to finish turning.

There is not much you can do with chopper headland. The above idea doesn't work 100% of the time due to game restraints unless they fixed it in 19. Maybe having the chopper share it's next few waypoints so it knows if the chopper is going to move away or closer to the tipper instead of waiting until after the fact this cause some TC or out range issues. This could also apply to combines.

And depending on the situation I use find on field. Early game I usually use manual so I don't have to remember to change. And the other auto detect doesn't work unless it is parked perfectly on the field.

I know it is long but I would like to see the old way of parking kept as both you and I spent alot of time trying to perfect it.

**Comment by KenStumich** [comment]
> Has someone that was trying to improve the old one. I know leaving the tipper where it is **doesn't work for field edge cases** as this interferes with the combines turn manuver and depending on the map it could also leave it parked in a roadway. There also could be issues with those when using multiple combines. Also having the tipper move gives people that have other tools working the field that extra space

the problem here is the edge. 

detect the ege and stop befor that edge to unload.  cp makes his own wayponits . if we can make a turnig point at waypoint x, we can also define a waypoint to stop at X- red zone.

define a rule where its safe to unload  and ask yourself how to detect the zone. 

![up-down pattern](https://user-images.githubusercontent.com/28400454/53289810-35961e80-379b-11e9-8249-71f2a7205284.png)

**Comment by Tankieboy** [comment]
Just gonna throw this in to the mix... If crop destruction is activated within the game settings then could we have crop destruction within courseplay? I ensure only tractors and equipment with narrow tyres enter the field to manage weeds and fertilise in the interests of realism. Strange seeing a tractor and trailor with normal/wide tyres drive through crop without consequence when unloading harvesters.

**Comment by pops64** [comment]
That is great for square fields but for diagonal it is a whole different situation. Also don't forget the tractor still as to turn around seeing how it is at the field.  That is why I liked the way it was done. It was simple but effective. Just need some of the fringe cases taken care of. The problem is Mode2 is so complex that trying to fix a fringe case usually breaks it

**Comment by mehdinasef** [comment]
I see that some great ideas can easily be applied to square fields but not to other shapes. I wonder why not to just make them available for square fields first since most maps come with square or rectangular fields

**Comment by GeorgeEarslight** [comment]
> > I'm only fluent in C
> 
> @GeorgeEarslight what a lame excuse :)

@pvaiko It was indeed quite lame :P
Here's an [idea](https://gist.github.com/GeorgeEarslight/5574b3398645420be69085c7e2d849ac)

**Comment by pvaiko** [comment]
@GeorgeEarslight thanks for the idea. One of the caveats of Courseplay development is the huge gap between having a general idea (10% of the time spent) and having a working implementation in the game (90% of the time spent).

**Comment by GeorgeEarslight** [comment]
> @GeorgeEarslight thanks for the idea. One of the caveats of Courseplay development is the huge gap between having a general idea (10% of the time spent) and having a working implementation in the game (90% of the time spent).

Isn't that the truth for most projects involving programming. 
Anyway, thanks for taking the time to take a look. Wish I get the time to study your code more thoroughly so I can test on my own.

**Comment by pvaiko** [comment]
> Isn't that the truth for most projects involving programming.

Somewhat yes, but even more in an environment where debugging and automated testing is practically non existent although we made some progress in this area lately.

For the current waypoint/course model you may want to look at Waypoint.lua and *AIDriver.lua for usage.

**Comment by Innallas** [comment]
You talk about a new combi mode where there will be a tractor who follow combi ? 
Why not using a variable sent by the combi who say "wait" to the tractor when the combi reach a turn-manoeuver node then he send him "you can come" when he reach the 2nd point where combi restart harvest > then the tractor start to turn , and the combi move forward like 10-20 meters to let place for the tractor to come ? Then when they are starting to load/unload they restart moving forward both together ?

**Comment by GeorgeEarslight** [comment]
> You talk about a new combi mode where there will be a tractor who follow combi ?

We were discussing about the tractor parking position. 
Thomas has some bullets points in the original  message, requesting feedback from the community.

>  so i'll try to help too :).
That's great! I'm sure that if you start trying to debug the issues here, you'll learn the CP code and get to know the developers too.

**Comment by Innallas** [comment]
I also was thinking about an easy combi mode or addition to one. It's easy and have to do :
- Field work (such as plowing/harvesting) 
> then merge/fusion a route who goes unload the trail/combi.

i'm taking straw or grass_dry with a loading wagon but he don't go unload with the route merged, is it possible to do ? maybe not so hard ? (like i can do maybe ?)

**Comment by Goof245** [comment]
Would be great for my use case if all combine reversing can be done with an articulated trailer, i.e. a power dolly with baler towed behind it... Failing that, options to disable anything that might reverse the combine would be greatly appreciated, like we could in FS17 by changing headland turn / turn on field options.

To answer the OP, 
>tractor next to combine, combine is empty -> where should we park the tractor ?

Parking behind the combine seemed to work well enough in FS17, I can't really think of any better way to do it.

>chopper is turning on  field edge, how do you imagine the turn maneuver of the tractor ?

Try to get behind the combine, aligned along the last path the combine cut. Then once in position, signal that the tractor is out of the way and it should be safe to start the turning maneuver. Maybe a failsafe could be a "go away" mode, if something's not going well, combine could tell the tractor to move right out of the way, tractor pathfinds through the cut crop and just keeps driving until the combine is happy and starts the turn... 

>who uses "search combine on field xx" already ?

Me :P 
Very handy on maps with tight fields, often the chaser bin parking bay isn't on the field at all. Being able to manually specify the field to search is very handy.

>any other remarks what should be changed ? now or never :-)

Combines and chaser bins trying to reap headland along irregular field edges can be hit and miss in FS17. Maybe try and have the tractor watch the edge of the header, and maintain a minimum distance so as not to trip the combine's traffic warning. Perhaps have the tractor stop and reverse back if the header is about to hit it, then resume unloading once the combine starts driving forward again?

**Comment by bgo1973** [comment]
Ich schreibe es mal auf deutsch. Das Vorgewende sollte nur an zwei gegenüber liegenden Seiten sein, nicht rings um das Feld herum. Normalerweise die kürzeren. Also z.B. zwei Bahnen Vorgewende, dann die erste "normale" Bahn gefolgt von den zwei Bahnen Vorgewende am anderen Ende des Feldes. Dann folgt das "normale auf und ab". Sollte der Drescher beim Abtanken das Ende einer Bahn erreichen, sollte er vor dem Wendemanöver zusammen mit dem Abfahrer halten, bis er leer ist. Der Abfahrer muss dann am Besten in die vorherige Spur linksherum wenden und hat somit eine Arbeitsbreite des Mähdreschers Abstand. Ist zum Zeitpunkt des eingestellten Abtankfüllstandes das Abtankrohr über der Frucht, sollte der Mähdrescher bei der Ankunft der Abfahrers stoppen und rückwärts eine Arbeitsbreite nach rechts fahren. Am Ende des Abtankens muss dann der Abfahrer warten, bis der Mähdrescher wieder in der aktuellen "Arbeitsspur" ist und wiederum eine Arbeitsbreite Platz schaffen.  So sollte es in den meisten Fällen nicht zu Kollisionen kommen. In Ausnahmefällen, wie z.B. auf der ersten Bahn nach dem ersten Vorgewende, müsste dann evtl. und wenn möglich vorberechnet werden, ob noch im Vorgewende, also bevor in die erste "normale" Bahn gewendet wird, abgetankt werden muss.

**Comment by bgo1973** [comment]
In english: The headland should only be on two opposite sides, not around the field. Usually the shorter ones. So e.g. two lanes headland, then the first "normal" track followed by the two lanes headland at the other end of the field. Then follows the "normal up and down." Should the combine reach the end of a track when overloading, it should stop together with the aidriver before the turning manoeuvre until it is empty. The aidriver then has to turn best into the previous track leftside and thus has a working width of the combine distance. If at the time of the adjusted filling stand the combines pipe is above the fruit, the combine should stop on the arrival of the aidriver and drive a working width to the right backwards. At the end of the overloading, the aidriver then has to wait until the combine is back in the current "work lane" and then has to turn right to create also a distance of one working width. In most cases, there should be caused no collisions. In exceptional cases, such as on the first row after the first headland, it might then have to be calculated and, if possible, whether it would have to be overloaded in the headland, i.e. before being turning into the first "normal" track.

**Comment by pvaiko** [comment]
> The headland should only be on two opposite sides, not around the field. Usually the shorter ones. So e.g. two lanes headland, then the first "normal" track followed by the two lanes headland at the other end of the field. Then follows the "normal up and down."

We have that already. Called 'short edges' headland mode ('Kurze Seiten')

**Comment by bgo1973** [comment]
I know. I only listed it to describe the eventual problem in the first row I described at the end of the post. Because maybe there will be not enough space on most fields to drive side by side. And if you harvest a complete round there will not be enough corntank volume. So maybe you will add this option (emptying after first headland) to the "short edges mode".

**Comment by bgo1973** [comment]
When harvesting with a harvester ( not combine) the aidriver should always drive behind the harvester on the complete headland. During the following "ups and downs" the harvester should use a Y-turn as the aidriver should use a wide omega-turn to avoid driving into the fruit and should then always drive on the side of the harvester turned away from the fruit.

**Comment by KenStumich** [comment]
why use an Y-turn , when the forage-harvester is able to do the U turn ? ( assuming your short edge headland. also if the unloading tractor is able to do the turn , why a "wide" omega turn ? priority shold be not to leave the field at all. the wide omega turn tends to collisions at the field border (e.g. fences or trees)

**Comment by bgo1973** [comment]
Maybe I'll do some pictures later. I think a U-turn needs more space at all.

**Comment by mehdinasef** [comment]
search combine on field xx is an amazing feature of CP. I always use it. But it can't be used on fields that I create using a plough which are not defined on the map, or combined fields where the combine can be detected on just one of them. in these situations, we record a course around the area or the created field and use it to generate the course for the harvester ("currently loaded course" option).
so can you make the overloader find the harvester in situations like these using the course that we recorded around the field?

**Comment by KenStumich** [comment]
> in these situations, we record a course around the area or the created field and use it to generate the course for the harvester ("currently loaded course" option).

why did you not use the automatic SCAN_CURRENT_FIELD_EDGES to scan the new field you cratet with the plow and save it as a new field ? a lot faster than create a manual fieldcourse by hand . you can use the new createt fields like the normal ones. search for some video about this or lookup in the wiki.

**Comment by mehdinasef** [comment]
Thank you, I did not know that exists. I have been using CP for quite a short time and don' know much about it.

**Comment by KenStumich** [comment]
no pro ..  have fun

**Comment by game455** [comment]
Combine full with pipe in fruit:
The maneuver works quite well, but the combine should drive a little further back, so that he is paralell to the course (with me he is always a little twisted). So the tractor may also come in a better loading position.
![fsScreen_2019_04_01_17_15_35](https://user-images.githubusercontent.com/47149345/55339087-4b85b600-54a2-11e9-95d3-29d5f1d3efea.png)
After the combine is empty, ** first ** the tractor should drive straight back (it can stop there too), so that the combine can then re-engage its track without getting in the way of the tractor.
![fsScreen_2019_04_01_17_20_34](https://user-images.githubusercontent.com/47149345/55339239-969fc900-54a2-11e9-84e1-47f7e9709328.png)
In this case, the tractor moves forward to the right, the combine simultaneously to the left.
Otherwise, a well-done extension 👍

**Comment by pvaiko** [comment]
@game455 the plan it to fine tune what the combine does during the mode 2 changes, so it isn't perfect yet, I know it should probably drive forward more before returning to work to avoid the tractor.

Not sure why is it twisted, I did not see that behavior, if you can post a screenshot or video with debug channel 14 while it is backing up would be great.

**Comment by game455** [comment]
ok. I'll do it on next field

**Comment by game455** [comment]
![fsScreen_2019_04_01_18_12_14](https://user-images.githubusercontent.com/47149345/55343730-fcdd1980-54ab-11e9-8e85-4b7ec9d34db8.png)
Combine drives straight back...
![fsScreen_2019_04_01_18_12_19](https://user-images.githubusercontent.com/47149345/55343782-141c0700-54ac-11e9-96eb-0f3001fae130.png)
...and just before the stop he steers fully to the left and stop
![fsScreen_2019_04_01_18_12_24](https://user-images.githubusercontent.com/47149345/55343812-28f89a80-54ac-11e9-81ae-0a6375688581.png)
[log.txt](https://github.com/Courseplay/courseplay/files/3030450/log.txt)
Backtrack manouver starts at 2019_04_01_18_12_07

**Comment by Wadeeb** [comment]
I am loving this maneuver where the combine pulls back out of the fruit! As a general comment, great work on Courseplay, it is a true game changer and having done a tiny amount of programming I appreciate the volume of work that goes into this.

**Comment by blubstar86** [comment]
since im most using CP on 4x or 16x maps and XXXL fields i would like to see additional speed controls for them on the field.

-when they drive from the point where they wait until the combine needs them to the the combine i would like them to drive full speed on the field and when they get closeer to the combine they should lower their speed step by step. Would also like to see an option to get always two or three drivers behind the combine, like the Forage Harverster do, most of the time its too late when the combine calls them and they have to drive long ways on the field.

-the additional speed control would also to be great for fieldwork (Balecolleting)

**Comment by TheAngryGamerTAG** [comment]
an idea for the combi mode where when the combine is full and it backs the outlet  pipe out of the fruit. maybe if the combine is like about 80 or 90% and there is a courseplay combi offload tractor within a certain range maybe have the tractor tell the combine to do that maneuver and have the tractor pause and wait for it to move to the position, then once the combine is empty have it wait until the combi offload tractor is clear or just for a few seconds before it proceeds to harvest. let me know your thought on this or if you have questions on what the hell that means.
